LightMetal - Add flatbuffer schema and conversion to/from functions for various types (#17039)

kmabeeTT opened the pull request:
### Ticket

I am breaking up original PR I put up recently into smaller bite-sized chunks as @omilyutin-tt suggested. Here is round 4.

> [Feature Request] Add Light Metal capture/replay initial changes to tt-metal for some workloads https://github.com/tenstorrent/tt-metal/issues/17039

### Problem description

This is initial/bootstrapping changes for "Light Metal" capture/replay feature that uses Flatbufers as serialization/deserialization library. See my previous bigger PR (https://github.com/tenstorrent/tt-metal/pull/16573) if you want to see how this will be used by followup merge, or context.

### What's changed

Changes from original PR, plus feedback items addressed.

 - Add the new flatbuffer schema files: `base_types.fbs  binary.fbs  buffer_types.fbs  command.fbs  program_types.fbs`
 - Add corresponding `to_flatbuffer()` and `from_flatbuffer()` conversion functions split up to match .fbs files in cpp/hpp files. These are what serialize C++ structs to flatbuffer binary during capture, and deserialize from flatbuffer binary back to C++ structs during reply.
 - cpp files are included in compile, but functions aren't used yet. Will be used after subsequent merge of light metal capture/replay libraries.
 - Make to/from_flatbuffer() friend of CircularBufferConfig to access internals
 - Remove default case statements from case statements for enums so when new enum values are added, compile error is seen to force updates.  
 - Remove inline v0 namespace that existed in the from_flatbuffer() files for no great reason.

### Checklist
- [x] Builds and docs passing
- [x] Post commit CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/13101964817)
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

kmabeeTT committed:
LightMetal - Add flatbuffer schema and conversion to/from functions for various types (#17039)

 - Code is compiled, but not used by anything yet, will be used after
   subsequent merge of light metal capture/replay libraries.
 - Remove default case statements from case statements for enums so when
   new enum values are added, compile error is seen to force updates.
 - All the PR feedback implemented. Add throws after switch statements for gcc12
 - CircularBufferConfig needed tweaks, add accessors for 3 private
   members for capture, and new constructor to set all private members
   for replay, following lengthy PR/offline discussion (avoid friend).

omilyutin-tt reviewed (changes requested):
Impressive! On high level just one concert around access to `CircularBufferConfig`.

omilyutin-tt commented on tt_metal/impl/flatbuffer/buffer_types_from_flatbuffer.cpp:
```diff
@@ -0,0 +1,61 @@
+        throw std::runtime_error("Invalid CircularBufferConfig FlatBuffer object");
+    }
+
+    // Create a CircularBufferConfig. Constructor doesn't matter much, since we serialized all
+    // members, will deserialize them here to get fully formed object.
+    CircularBufferConfig config(0, {});
```
I don't think we need to expose all of the stuff as public, right? A lot is already available. I'm worries about future changes yep, but also that we are re-creating `CircularBufferConfig` piece-by-piece instead of relying on the public API (ctor), and supplying the necessary information through `set*` methods. We are at risk of breaking invariants, if one data member depends on the other in some implicit way that is enforced by correct usage of the public APIs.

nsmithtt replied:
From a serialization perspective it looks like we're missing three getters/setters:
- `dynamic_cb_`
- `max_size_`
- `buffer_size_`

Are we ok with making those public API?

And then from a deserialization perspective , it's a little bit awkward because we'll have to always loop `NUM_CIRCULAR_BUFFERS` times for both local and remote buffers and get `Builder`s for them in order to set data formats, but not for the others.  It might be better (and more efficient in terms of deserialization) to have a new constructor that just forwards all of the members:

```c++
CircularBufferConfig(uint32_t total_size,         
  std::optional<uint32_t> globally_allocated_address,                  
  std::array<std::optional<tt::DataFormat>, NUM_CIRCULAR_BUFFERS> data_formats,   
  std::array<std::optional<uint32_t>, NUM_CIRCULAR_BUFFERS> page_sizes,   
  std::array<std::optional<Tile>, NUM_CIRCULAR_BUFFERS> tiles,   
  std::unordered_set<uint8_t> buffer_indices,   
  std::unordered_set<uint8_t> local_buffer_indices,   
  std::unordered_set<uint8_t> remote_buffer_indices,   
  bool dynamic_cb,                  
  uint32_t max_size,                            
  uint32_t buffer_size),
```

kmabeeTT replied:
If we don't have resolution here, let me know if we're open to deferring/revisiting this topic of "can we serialize/deserialize classes using private members, or must expose everything to be public" for later, to unblock this and the following changes I am trying to merge. This is all just round 1 so I can check in what I have working since early Dec and continue to improve it.

omilyutin-tt replied:
Yeah, this isn't ideal... The comment above regarding `Tile` is the same in spirit, but much more straighforward to deal with :)

We (in Metal) need to take a look at these members more closely. I see they exist for some validation that only gets triggered in debug mode. I see there is a TODO to clean these up actually @tt-aho, could you provide some context here? Do you think these members can be deleted now (with cleaned up tests)? Do you think we can serialize `CircularBufferConfig` in flatbuffers and just ignore these 3 data members, then deserialize and set them to some default values?

I also see there is `operator==` defined in the following way:
```
bool operator==(const CircularBufferConfig& lhs, const CircularBufferConfig& rhs) {
    return lhs.total_size() == rhs.total_size() &&
           lhs.globally_allocated_address() == rhs.globally_allocated_address() &&
           lhs.data_formats() == rhs.data_formats() && lhs.page_sizes() == rhs.page_sizes() &&
           lhs.tiles() == rhs.tiles() && lhs.shadow_global_buffer == rhs.shadow_global_buffer;
}
```
So it simply ignores some members.

@kmabeeTT @nsmithtt if we can ignore these, that would be amazing. Adding a new constructor (or perhaps using some of the existing setters?) also sounds good. What I am trying to preserve here is the "constructor as the public creation API". E.g. if someone decides tomorrow to add another data member - this is fine, as long as we are relying on the "official way" of creating `CircularBufferConfig`. If ctor changes - the dev can update the deserialization code too. Things will be a lot more fragile with friend functions.

kmabeeTT replied:
Currently don't think it's safe to ignore these 3 members since I am serializing/deserializing fully formed `CircularBufferConfig` at the time of usage (in `CreateCircularBuffer()` API call), so after some offline discussion I went ahead and created the new constructor and 3 getters, and updated my code to remove friend relationship, and use public APIs only now (getters and constructor) as requested.

```
    // For flatbuffer deserialization, set all private members.
    CircularBufferConfig(
        uint32_t total_size,
        std::optional<uint32_t> globally_allocated_address,
        const std::array<std::optional<tt::DataFormat>, NUM_CIRCULAR_BUFFERS>& data_formats,
        const std::array<std::optional<uint32_t>, NUM_CIRCULAR_BUFFERS>& page_sizes,
        const std::array<std::optional<Tile>, NUM_CIRCULAR_BUFFERS>& tiles,
        const std::unordered_set<uint8_t>& buffer_indices,
        const std::unordered_set<uint8_t>& local_buffer_indices,
        const std::unordered_set<uint8_t>& remote_buffer_indices,
        bool dynamic_cb,
        uint32_t max_size,
        uint32_t buffer_size);
        
<snip>

    // These 3 getters are not typically used, but needed for flatbuffer serialization
    // in order to avoid frend relationship. Added to constructor too, for deserialization.
    bool dynamic_cb() const;
    uint32_t max_size() const;
    uint32_t buffer_size() const;
    
```

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types.fbs:
```diff
@@ -0,0 +1,102 @@
+
+
+table Tile {
+  tile_shape: [uint32]; // Shape of the tile (e.g., height, width)
```
I mean, this is similar situation, where my goal was to snapshot all members and reconstruct from members.  I guess we don't like that. I'll have to look into this suggestion.

omilyutin-tt replied:
You are right it is a very similar situation as the circular buffer, but here I think it is much easier to rely on the constructor. Basically, constructor = stable public API to create a `Tile`, and if anyone messes with this, they will be forced to update the serialization routines.

kmabeeTT replied:
Made update, serialization/deserialization is just a couple lines each now.  Added assert for 2D size on deserialization from vector back to array as suggested below.

```
--- a/tt_metal/impl/flatbuffer/base_types_to_flatbuffer.cpp
+++ b/tt_metal/impl/flatbuffer/base_types_to_flatbuffer.cpp
@@ -85,20 +85,11 @@ tt::tt_metal::flatbuffer::DataFormat to_flatbuffer(tt::DataFormat input) {

 flatbuffers::Offset<tt::tt_metal::flatbuffer::Tile> to_flatbuffer(
     const Tile& tile, flatbuffers::FlatBufferBuilder& builder) {
-    auto tile_shape_fb = builder.CreateVector(tile.get_tile_shape().data(), tile.get_tile_shape().size());
-    auto face_shape_fb = builder.CreateVector(tile.get_face_shape().data(), tile.get_face_shape().size());

     return tt::tt_metal::flatbuffer::CreateTile(
         builder,
-        tile_shape_fb,
-        face_shape_fb,
-        tile.get_tile_hw(),
-        tile.get_face_hw(),
-        tile.get_num_faces(),
-        tile.get_partial_face(),
-        tile.get_narrow_tile(),
-        tile.get_transpose_within_face(),
-        tile.get_transpose_of_faces());
+        builder.CreateVector(tile.get_tile_shape().data(), tile.get_tile_shape().size()),
+        tile.get_transpose_within_face() && tile.get_transpose_of_faces());
 }
 ```

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types.fbs:
```diff
@@ -0,0 +1,102 @@
+  DM_DYNAMIC_NOC
+}
+
+enum Eth : ubyte {
```
Can we leave name cleanup for another day/task? My goal here was to adhere to same names of structs whenever possible and not create my own, no mattery how non-ideal they are.

omilyutin-tt replied:
Thanks! Yeah it could have been a follow up. Unfortunately a lot of these issues are in the existing code, and it is really great that we are incrementally making it better.

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types.fbs:
```diff
@@ -0,0 +1,102 @@
+
+
+table Tile {
+  tile_shape: [uint32]; // Shape of the tile (e.g., height, width)
```
I am just following the same form the Tile class uses (an array, vector in flatbuffer since arrays not supported), and preferred not to get creative with some new container.  Maybe one day they'll change this to 3D shape...

```
std::array<uint32_t, 2> tile_shape = {constants::TILE_HEIGHT, constants::TILE_WIDTH};
```

omilyutin-tt replied:
Sure, not a strong opinion on this. But also to say `std::array<int, 2>` is more like a fixed size than a dynamically sized vector. We might want to add an assertion that the serialized length of the data is 2?

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types_from_flatbuffer.hpp:
```diff
@@ -0,0 +1,34 @@
+
+Tile from_flatbuffer(const tt::tt_metal::flatbuffer::Tile* tile_fb);
+
+std::array<std::optional<Tile>, NUM_CIRCULAR_BUFFERS> from_flatbuffer(
```
Don't think I have any choice here, Tile has an element of fixed size array, `std::array<std::optional<Tile>, NUM_CIRCULAR_BUFFERS> ` so this is me populating it.  Flatbuffers don't have arrays, only vectors.

omilyutin-tt replied:
I think we need to add assertion then that the number of elements in the serialized flatbuffer would equal to `NUM_CIRCULAR_BUFFERS`?

kmabeeTT replied:
Thanks - added assert, this helped me catch a bug. I was skipping pushing nullopt Tiles into the vector during serialization, leading to fewer than NUM_CIRCULAR_BUFFERS entries during deserialization. Whoops.

kmabeeTT replied:
Many hours on this to find that vector of Tables cannot have null values (despite what AI says), so I had to fix this to make Tiles vector like the other vectors of optional elements by having an intermediate type to represent a KV pair of which index in original C++ optional array this Tile is from.  

```

struct CBConfigPageSize {
  index: uint32;  // The index in the array
  size: uint32;   // The page-size value for this index
}

struct CBConfigDataFormat {
  index: uint32;  // The index in the array
  format: DataFormat; // The data format for this index
}


struct CBConfigTile {
  index: uint32;  // The index in the array
  tile: Tile;     // The tile for this index
}

table CircularBufferConfig {
...
  data_formats: [CBConfigDataFormat]; // Mimic optional array in C++ by using KV map.
  page_sizes: [CBConfigPageSize];     // Mimic optional array in C++ by using KV map.
  tiles: [CBConfigTile];              // Mimic optional array in C++ by using KV map.
  ```

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types_to_flatbuffer.hpp:
```diff
@@ -0,0 +1,36 @@
+    const Tile& tile, flatbuffers::FlatBufferBuilder& builder);
+
+flatbuffers::Offset<flatbuffers::Vector<flatbuffers::Offset<tt::tt_metal::flatbuffer::Tile>>> to_flatbuffer(
+    const std::array<std::optional<Tile>, NUM_CIRCULAR_BUFFERS>& tiles, flatbuffers::FlatBufferBuilder& builder);
```
Don't see a point, because the thing I am trying to serialize is a `std::array<std::optional<Tile>, NUM_CIRCULAR_BUFFERS>& tiles` , I have to take the types as they exist on structs/classes.

omilyutin-tt replied:
Ok, I think other comments make this case more clear.

kmabeeTT replied:
Thx, will resolve this one. Chipping away, hope to push something once all addressed and tested.

kmabeeTT commented on tt_metal/impl/flatbuffer/program_types.fbs:
```diff
@@ -0,0 +1,74 @@
+
+namespace tt.tt_metal.flatbuffer;
+
+table CoreCoord {
+  x: int;
+  y: int;
+}
+
+table CoreRange {
+  start: CoreCoord;
+  end: CoreCoord;
+}
```
Sure that is fair, I can change to struct.

kmabeeTT replied:
Unfortunately cannot mix Structs and Tables in the same Union, and like C++ we treat `CoreCoord`, `CoreRange` and `CoreRanges` (where `CoreRanges` is dynamic length vector, must be table) as a union. So I have to leave this alone.

Did convert these to structs though (which required updating code that uses them):

```
struct UInt32Value {
  value: uint32;
}

struct BufferGlobalId {
  id: uint32;
}
```

kmabeeTT commented on tt_metal/impl/flatbuffer/buffer_types.fbs:
I really liked the types suffix (echo my previous comments on parent PR) because it's not really a buffer, and it's not a buffer config (another term already used), but rather a (rather arbitrary, best effort I did) collection of all the various types (a term used in our hpp files often to contain these definitions for enums, structs) that are related or tangentially related to buffers...

omilyutin-tt replied:
Maybe this is my allergy to the 1980s C style headers:)

I see your point. Let's re-evaluate down the road, if necessary.

kmabeeTT commented on tt_metal/impl/flatbuffer/program_types_from_flatbuffer.cpp:
```diff
@@ -0,0 +1,143 @@
+namespace tt::tt_metal {
+
+std::variant<CoreCoord, CoreRange, CoreRangeSet> from_flatbuffer(
+    const tt::tt_metal::flatbuffer::CoreSpec core_spec, const void* flatbuffer_union) {
```
Good point. I was going to say No, but then I found there is an optional flag to "flatc" generation of C++ headers to _apparently_ generate some type safe conversions for unions. Not currently enabled, will try.  It's called `--gen-object-api`

We call it this way today (this will change with above idea):

```
auto core_spec = from_flatbuffer(cmd->core_spec_type(), cmd->core_spec());
```

kmabeeTT replied:
Turns out these `as_*()` functions are pretty much the same (they use `static_cast` as I was doing). Lets look at `KernelConfig` as example (same thing, smaller code than `CoreSpec`):

```
  const tt::tt_metal::flatbuffer::DataMovementConfig *kernel_config_as_DataMovementConfig() const {
    return kernel_config_type() == tt::tt_metal::flatbuffer::KernelConfig::DataMovementConfig ? static_cast<const tt::tt_metal::flatbuffer::DataMovementConfig *>(kernel_config()) : nullptr;
  }
```

It's a bit nicer in that it handles the `static_cast` internally, though same net effect. The downside if you need to pass parent table to `from_flatbuffer()` function, which means the all commands must have `CoreSpec` or `KernelConfig` elements named the same.  Anyways, for now I made the change since it took a long time to figure this out, can be used as a reference, and will leave it as a separate commit in case we want to revert it later.

Before (in `program_types_from_flatbuffer.cpp`) we had:

```
std::variant<DataMovementConfig, ComputeConfig, EthernetConfig> from_flatbuffer(
    const flatbuffer::KernelConfig config_type, const void* flatbuffer_union) {
    switch (config_type) {
        case flatbuffer::KernelConfig::DataMovementConfig:
            return from_flatbuffer(static_cast<const flatbuffer::DataMovementConfig*>(flatbuffer_union));
        case flatbuffer::KernelConfig::ComputeConfig:
            return from_flatbuffer(static_cast<const flatbuffer::ComputeConfig*>(flatbuffer_union));
        case flatbuffer::KernelConfig::EthernetConfig:
            return from_flatbuffer(static_cast<const flatbuffer::EthernetConfig*>(flatbuffer_union));
        case flatbuffer::KernelConfig::NONE:
            throw std::runtime_error("Unhandled KernelConfig type in from_flatbuffer.");
    }
    TT_THROW("Unhandled KernelConfig type in from_flatbuffer.");
}
```

Now (in `program_types_from_flatbuffer.hpp`) we have:

```
template <typename CommandType>
std::variant<DataMovementConfig, ComputeConfig, EthernetConfig> kernel_config_from_flatbuffer(const CommandType* cmd) {
    switch (cmd->kernel_config_type()) {
        case flatbuffer::KernelConfig::DataMovementConfig:
            return from_flatbuffer(cmd->kernel_config_as_DataMovementConfig());
        case flatbuffer::KernelConfig::ComputeConfig: 
            return from_flatbuffer(cmd->kernel_config_as_ComputeConfig());
        case flatbuffer::KernelConfig::EthernetConfig: 
            return from_flatbuffer(cmd->kernel_config_as_EthernetConfig());
        case flatbuffer::KernelConfig::NONE:
            throw std::runtime_error("Unhandled KernelConfig type in from_flatbuffer.");
    }
    TT_THROW("Unhandled KernelConfig type in from_flatbuffer.");
}
```

omilyutin-tt reviewed (changes requested):
Btw, do we want to add "Roundtrip" tests for these serialization / deserialization routines (something like `EXPECT_EQ(from_flatbuffer(to_flatbuffer(object)), object)`)? Whatever makes sense at the top level - for complex objects like the circular buffer config it might be important to capture the behavior.

kmabeeTT commented:
> Btw, do we want to add "Roundtrip" tests for these serialization / deserialization routines (something like `EXPECT_EQ(from_flatbuffer(to_flatbuffer(object)), object)`)? Whatever makes sense at the top level - for complex objects like the circular buffer config it might be important to capture the behavior.

Yeah, I like that idea - thanks for flagging it!  I will add this to list of things to do after current "set" of changes from my original PR are in (these are just initial changes to bootstrap the feature, not intended to be perfect).

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types_from_flatbuffer.hpp:
```diff
@@ -0,0 +1,34 @@
+
+BufferType from_flatbuffer(tt::tt_metal::flatbuffer::BufferType type);
+
+tt::tt_metal::DataMovementProcessor from_flatbuffer(tt::tt_metal::flatbuffer::DataMovementProcessor in);
```
Good point, it is true. And even the non-return types like `tt::tt_metal::flatbuffer::DataMovementProcessor` here. I removed all tt::tt_metal:: references to make the code tighter, bunch of things clang-format collapsed to single lines now, nice.

kmabeeTT commented on tt_metal/impl/flatbuffer/program_types.fbs:
Keeping this for now, similar discussion below https://github.com/tenstorrent/tt-metal/pull/17353#discussion_r1936630157

kmabeeTT commented on tt_metal/impl/flatbuffer/program_types_from_flatbuffer.hpp:
```diff
@@ -0,0 +1,26 @@
+std::variant<DataMovementConfig, ComputeConfig, EthernetConfig> from_flatbuffer(
+    const tt::tt_metal::flatbuffer::KernelConfig config_type, const void* flatbuffer_union);
+
+tt::stl::Span<const SubDeviceId> from_flatbuffer(const flatbuffers::Vector<uint8_t>* fb_sub_device_ids);
```
Damn, good catch... agreed, fixed it, and use implicit conversion from vector to span on the usage of the returned vector on call to Finish() API.

kmabeeTT commented on tt_metal/impl/flatbuffer/program_types_to_flatbuffer.cpp:
```diff
@@ -0,0 +1,211 @@
+                auto ranges_vector = builder.CreateVector(range_offsets);
+                auto core_range_set = tt::tt_metal::flatbuffer::CreateCoreRangeSet(builder, ranges_vector);
+                return {tt::tt_metal::flatbuffer::CoreSpec::CoreRangeSet, core_range_set.Union()};
+            } else {
+                throw std::runtime_error("Unhandled variant type in to_flatbuffer");
```
Nice, limited experience with this type of coding, but I tried it out and seems to work, made the change here and one other place.

kmabeeTT commented on tt_metal/impl/flatbuffer/program_types_to_flatbuffer.cpp:
```diff
@@ -0,0 +1,211 @@
+        value_offset = buffer_offset.Union();
+        value_type = tt::tt_metal::flatbuffer::RuntimeArgValue::BufferGlobalId;
+    } else {
+        throw std::runtime_error("Unexpected variant type in createRuntimeArg");
```
Made that chage.  Then made the change to update this to use tt::stl::overloaded, which I think if I'm understanding it correctly means I don't need the static_assert anymore since it ensures all variants are supported.

kmabeeTT commented:
FYI @omilyutin-tt  - I force pushed a commit (will fixup before merge) to address all feedback (except the not-quite closed `CircularBufferConfig` public vs private discussion), including rebase on a recently close-to-passing commit.   Can I get approval or re-review?  I will still address the`CircularBufferconfig` topic still before merging (tonight or tomorrow) by adding new public accessors for the missing private elements during serialization, and constructor for deserialization but it will take a bit longer and I have to head out now.  Then I'll start on carving off the next set of changes for the next PR after getting this in.

kmabeeTT committed:
LightMetal - Use as_*() instead of static_cast directly for some union type handling in from_flatbuffer()

 - CoreSpec and KernelConfig are unions, do need to use static_cast
   but for now will prefer to hide that detail by using
   cmd->core_spec_as_CoreCoord() instead.
 - Downside here is that field names core_spec and kernel_config must be
   consistent across all users of these from_flatbuffer() funcs now.
